README.md
# react-native-custom-datetimepicker [![Build Status](https://travis-ci.org/xgfe/react-native-custom-datetimepicker.svg?branch=master)](https://travis-ci.org/xgfe/react-native-custom-datetimepicker) [![Coverage Status](https://coveralls.io/repos/github/xgfe/react-native-custom-datetimepicker/badge.svg?branch=master)](https://coveralls.io/github/xgfe/react-native-custom-datetimepicker?branch=master) [![Monthly download](https://img.shields.io/npm/dm/react-native-custom-datetimepicker.svg)](https://img.shields.io/npm/dm/react-native-custom-datetimepicker.svg) [![Total downloads](https://img.shields.io/npm/dt/react-native-custom-datetimepicker.svg)](https://img.shields.io/npm/dt/react-native-custom-datetimepicker.svg)
react native datePicker component for both Android and IOS, using DatePickerAndroid, TimePickerAndroid and DatePickerIOS

## Install

```bash
npm install react-native-custom-datetimepicker --save
```

Or using [react-native-ui-xg](https://github.com/xgfe/react-native-ui-xg), our react-naitve ui kit.

```bash
npm install react-native-ui-xg --save
```

## Example
Check [index.android.js](https://github.com/xgfe/react-native-custom-datetimepicker/blob/master/index.android.js) in the Example.

![android](http://xgfe.github.io/react-native-custom-datetimepicker/img/react-native-custom-datetimepicker-android.gif)
![ios](http://xgfe.github.io/react-native-custom-datetimepicker/img/react-native-custom-datetimepicker-ios.gif)

## Usage

```javascript

import React, { Component } from 'react'
import DatePicker from 'react-native-custom-datetimepicker'

export default class MyDatePicker extends Component {
  constructor(props){
    super(props)
    this.state = {date:"2016-05-15"}
  }

  render(){
    return (
      <DatePicker
        style={{width: 200}}
        date={this.state.date}
        mode="date"
        placeholder="select date"
        format="YYYY-MM-DD"
        minDate="2016-05-01"
        maxDate="2016-06-01"
        confirmBtnText="Confirm"
        cancelBtnText="Cancel"
        customStyles={{
          dateIcon: {
            position: 'absolute',
            left: 0,
            top: 4,
            marginLeft: 0
          },
          dateInput: {
            marginLeft: 36
          }
          // ... You can check the source to find the other keys.
        }}
        onDateChange={(date) => {this.setState({date: date})}}
      />
    )
  }
}
```

You can check [index.js](https://github.com/xgfe/react-native-custom-datetimepicker/blob/master/index.android.js) in the Example for detail.

## Properties

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| style | - | `object` | Specify the style of the DatePicker, eg. width, height...  |
| date | - | <code>string &#124; date &#124; Moment instance</code> | Specify the display date of DatePicker. `string` type value must match the specified format |
| mode | 'date' | `enum` | The `enum` of `date`, `datetime` and `time` |
| androidMode | 'default' | `enum` | The `enum` of `default`, `calendar` and `spinner` (only Android) |
| format | 'YYYY-MM-DD' | `string` | Specify the display format of the date, which using [moment.js](http://momentjs.com/). The default value change according to the mode. |
| confirmBtnText | '确定' | `string` | Specify the text of confirm btn in ios. |
| cancelBtnText | '取消' | `string` | Specify the text of cancel btn in ios. |
| iconSource | - | <code>{uri: string} &#124; number</code> | Specify the icon. Same as the `source` of Image, always using `require()` |
| minDate | - | <code>string &#124; date</code> | Restricts the range of possible date values. |
| maxDate | - | <code>string &#124; date</code> | Restricts the range of possible date values. |
| duration | 300 | `number` | Specify the animation duration of datepicker.|
| customStyles | - | `number` | The hook of customize datepicker style, same as the native style. `dateTouchBody`, `dateInput`...|
| showIcon | true | `boolean` | Controller whether or not show the icon |
| hideText | false | `boolean` | Controller whether or not show the `dateText` |
| iconComponent | - | `element` | Set the custom icon |
| disabled | false | `boolean` | Controller whether or not disable the picker |
| is24Hour | - | `boolean` | Set the TimePicker is24Hour flag. The default value depend on `format`. Only work in Android |
| placeholder | '' | `string` | The placeholder show when this.props.date is falsy |
| onDateChange | - | `function` | This is called when the user confirm the picked date or time in the UI. The first and only argument is a date or time string representing the new date and time formatted by [moment.js](http://momentjs.com/) with the given format property. |
| onOpenModal | - | `function` | This is called when the DatePicker Modal open. |
| onCloseModal | - | `function` | This is called when the DatePicker Modal close |
| onPressMask | - | `function` | This is called when clicking the ios modal mask |
| modalOnResponderTerminationRequest | - | `function` | Set the callback for React Native's [Gesture Responder System](https://facebook.github.io/react-native/docs/gesture-responder-system.html#responder-lifecycle)'s call to `onResponderTerminationRequest`. By default this will reject a termination request, but can be overidden in case the View under the Modal is implementing custom gesture responders, and you wish for those to be overidden in certain cases.  |
| TouchableComponent | `TouchableHighlight` | `Component` | Replace the `TouchableHighlight` with a custom `Component`. For example : `TouchableOpacity` |

### Property `customStyles` available keys

* appearance: `dateInput`, `disabled`, `dateTouchBody`, `dateIcon`, `placeholderText`, `dateText`
* ios select panel: `datePickerCon`, `datePicker`, `btnConfirm`, `btnTextConfirm`, `btnCancel`, `btnTextCancel`


## Instance Methods

| Method  | Params  | Description |
| :------------ |:---------------:| :---------------:|
| onPressDate | - | Manually open the date picker panel |
| onPressCancel | - | Manually close the date picker panel like, similarly pressing cancel btn |


CHANGELOG.md
<a name="1.6.0"></a>
# [1.6.0](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.5.1...v1.6.0) (2017-06-12)

### Features

* **datepicker:** adding testIds to elements ([#138](https://github.com/xgfe/react-native-custom-datetimepicker/issues/138)
* **datepicker:** add `hideText` property to hide the Date Text ([#127](https://github.com/xgfe/react-native-custom-datetimepicker/issues/127)
* **datepicker:** add `TouchableComponent` property to support custom touchable components ([#128](https://github.com/xgfe/react-native-custom-datetimepicker/issues/128)


<a name="1.5.1"></a>
## [1.5.1](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.5.0...v1.5.1) (2017-05-16)


### Features

* **datepicker:** change datepicker switch gap timeout to 200ms ([5b6e78f](https://github.com/xgfe/react-native-custom-datetimepicker/commit/5b6e78f))

### Bug Fixes

* **datepicker:** Fix onDateChange for ios ([#123](https://github.com/xgfe/react-native-custom-datetimepicker/issues/123), [e4e2193](https://github.com/xgfe/react-native-custom-datetimepicker/commit/e4e2193))


<a name="1.5.0"></a>
# [1.5.0](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.4.7...v1.5.0) (2017-05-15)


### Features

* **datepicker:** Add onPressMask prop for ios ([#120](https://github.com/xgfe/react-native-custom-datetimepicker/issues/120), [09ecdf8](https://github.com/xgfe/react-native-custom-datetimepicker/commit/09ecdf8))
* **datepicker:** Ability to pass icon component with `iconComponent` property ([#124](https://github.com/xgfe/react-native-custom-datetimepicker/issues/124), [35150ed](https://github.com/xgfe/react-native-custom-datetimepicker/commit/35150ed))

### Bug Fixes

* **datepicker:** Fix onCloseModal for android ([#122](https://github.com/xgfe/react-native-custom-datetimepicker/issues/122), [ba55c06](https://github.com/xgfe/react-native-custom-datetimepicker/commit/ba55c06))


<a name="1.4.7"></a>
## [1.4.7](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.4.6...v1.4.7) (2017-04-13)


### Bug Fixes

* **datepicker:** fix `androidMode` prop pass bug ([#111](https://github.com/xgfe/react-native-custom-datetimepicker/issues/111), [095f756](https://github.com/xgfe/react-native-custom-datetimepicker/commit/095f756))


<a name="1.4.6"></a>
## [1.4.6](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.4.5...v1.4.6) (2017-04-05)


### Features

* **datepicker:** add `androidMode` property for DatePickerAndroid & TimePickerAndroid ([be30731](https://github.com/xgfe/react-native-custom-datetimepicker/commit/be30731))



<a name="1.4.5"></a>
## [1.4.5](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.4.4...v1.4.5) (2017-03-21)


### Bug Fixes

* **datepicker:** fix `date` prop change not invoke state change bug ([#93](https://github.com/xgfe/react-native-custom-datetimepicker/issues/93), [3c8099b](https://github.com/xgfe/react-native-custom-datetimepicker/commit/3c8099b))



<a name="1.4.4"></a>
## [1.4.4](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.4.3...v1.4.4) (2017-02-06)


### Features

* **datepicker:** Added two new properties: onOpenModal and onCloseModal ([#77](https://github.com/xgfe/react-native-custom-datetimepicker/pull/77), [846ae35](https://github.com/xgfe/react-native-custom-datetimepicker/commit/846ae35))
* **Datepicker:** Add animation on datepicker dismiss ([#76](https://github.com/xgfe/react-native-custom-datetimepicker/pull/76), [fdfcc21](https://github.com/xgfe/react-native-custom-datetimepicker/commit/fdfcc21))

### Bug Fixes

* **datepicker:** fix datepicker can't  open in landscape mode bug ([#28](https://github.com/xgfe/react-native-custom-datetimepicker/issues/28), [3a9df99](https://github.com/xgfe/react-native-custom-datetimepicker/commit/3a9df99))



<a name="1.4.3"></a>
## [1.4.3](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.4.1...v1.4.3) (2016-12-13)


### Bug Fixes

* **datepicker:** fix `disabled` props change not update bug [#66](https://github.com/xgfe/react-native-custom-datetimepicker/issues/66) ([7579dc9](https://github.com/xgfe/react-native-custom-datetimepicker/commit/7579dc9)), closes [#66](https://github.com/xgfe/react-native-custom-datetimepicker/issues/66)
* **datepicker:** fix is24Hour flag can't be specified bug ([c320e56](https://github.com/xgfe/react-native-custom-datetimepicker/commit/c320e56)), closes [#69](https://github.com/xgfe/react-native-custom-datetimepicker/issues/69)


### Reverts

* **datepicker:** revert responder prevert func temporary [#47](https://github.com/xgfe/react-native-custom-datetimepicker/issues/47) [#50](https://github.com/xgfe/react-native-custom-datetimepicker/issues/50) ([5b9f60c](https://github.com/xgfe/react-native-custom-datetimepicker/commit/5b9f60c))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.4.0...v1.4.1) (2016-11-03)


### Bug Fixes

* **datepicker:** fix btn not work in 7p bug [#50](https://github.com/xgfe/react-native-custom-datetimepicker/issues/50) ([8285d2d](https://github.com/xgfe/react-native-custom-datetimepicker/commit/8285d2d)), closes [#50](https://github.com/xgfe/react-native-custom-datetimepicker/issues/50)



<a name="1.4.0"></a>
# [1.4.0](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.3.2...v1.4.0) (2016-10-31)


### Features

* **datepicker:** default prevent manual panResonder when mdal shown ([6d1c1d0](https://github.com/xgfe/react-native-custom-datetimepicker/commit/6d1c1d0))
* **Datepicker:** support minuteInterval property in ios [#46](https://github.com/xgfe/react-native-custom-datetimepicker/issues/46) ([3ae69a4](https://github.com/xgfe/react-native-custom-datetimepicker/commit/3ae69a4))



<a name="1.3.2"></a>
## [1.3.2](https://github.com/xgfe/react-native-custom-datetimepicker/compare/v1.3.1...v1.3.2) (2016-08-02)


### Bug Fixes


style.js
import { StyleSheet } from 'react-native';

let style = StyleSheet.create({
  dateTouch: {
    width: "100%",
    borderBottomWidth: 1,
  },
  dateTouchBody: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  dateIcon: {
    width: 22,
    height: 22,
    marginLeft: 5,
  },
  dateInput: {
    flex: 1,
    height: 40,
    borderWidth: 0,
    borderColor: '#aaa',
    alignItems: 'flex-start',
    justifyContent: 'center'
  },
  dateText: {
    color: '#333'
  },
  placeholderText: {
    color: '#c9c9c9'
  },
  datePickerMask: {
    flex: 1,
    alignItems: 'flex-end',
    flexDirection: 'row',
    backgroundColor: '#00000077'
  },
  datePickerCon: {
    backgroundColor: '#fff',
    height: 0,
    overflow: 'hidden'
  },
  btnText: {
    position: 'absolute',
    top: 0,
    height: 42,
    padding: 20,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  btnTextText: {
    fontSize: 16,
    color: '#46cf98'
  },
  btnTextCancel: {
    color: '#666'
  },
  btnCancel: {
    left: 0
  },
  btnConfirm: {
    right: 0
  },
  datePicker: {
    marginTop: 42,
    borderTopColor: '#ccc',
    borderTopWidth: 1
  },
  disabled: {
    backgroundColor: '#eee'
  }
});

export default style;


index.js
import React, {Component} from 'react';
import {
  AppRegistry,
  StyleSheet,
  Text,
  View,
  PanResponder
} from 'react-native';
import DatePicker from './datepicker.js';

class datepicker extends Component {

  constructor(props) {
    super(props);

    this.state = {
      date: '',
      time: '20:00',
      datetime: '2016-05-05 20:00',
      datetime1: '2016-05-05 20:00'
    };
  }

  componentWillMount() {
    this._panResponder = PanResponder.create({
      onStartShouldSetPanResponder: (e) => {console.log('onStartShouldSetPanResponder'); return true;},
      onMoveShouldSetPanResponder: (e) => {console.log('onMoveShouldSetPanResponder'); return true;},
      onPanResponderGrant: (e) => console.log('onPanResponderGrant'),
      onPanResponderMove: (e) => console.log('onPanResponderMove'),
      onPanResponderRelease: (e) => console.log('onPanResponderRelease'),
      onPanResponderTerminate: (e) => console.log('onPanResponderTerminate')
    });
  }

  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.welcome}>
          Welcome to react-native-custom-datetimepicker example!
        </Text>
        <DatePicker
          style={{width: 200}}
          date={this.state.date}
          mode="date"
          placeholder="placeholder"
          format="YYYY-MM-DD"
          minDate="2016-05-01"
          maxDate="2016-06-01"
          confirmBtnText="Confirm"
          cancelBtnText="Cancel"
          iconSource={require('./google_calendar.png')}
          onDateChange={(date) => {this.setState({date: date});}}
        />
        <Text style={styles.instructions}>date: {this.state.date}</Text>
        <DatePicker
          style={{width: 200}}
          date={this.state.time}
          mode="time"
          format="HH:mm"
          confirmBtnText="Confirm"
          cancelBtnText="Cancel"
          minuteInterval={10}
          onDateChange={(time) => {this.setState({time: time});}}
        />
        <Text style={styles.instructions}>time: {this.state.time}</Text>
        <DatePicker
          style={{width: 200}}
          date={this.state.datetime}
          mode="datetime"
          format="YYYY-MM-DD HH:mm"
          confirmBtnText="Confirm"
          cancelBtnText="Cancel"
          showIcon={false}
          onDateChange={(datetime) => {this.setState({datetime: datetime});}}
        />
        <Text style={styles.instructions}>datetime: {this.state.datetime}</Text>
        <DatePicker
          style={{width: 200}}
          date={this.state.datetime1}
          mode="datetime"
          format="YYYY-MM-DD HH:mm"
          confirmBtnText="Confirm"
          cancelBtnText="Cancel"
          customStyles={{
            dateIcon: {
              position: 'absolute',
              left: 0,
              top: 4,
              marginLeft: 0
            },
            dateInput: {
              marginLeft: 36
            }
          }}
          minuteInterval={10}
          onDateChange={(datetime) => {this.setState({datetime1: datetime});}}
        />
        <Text style={styles.instructions}>datetime: {this.state.datetime1}</Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF'
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5
  }
});

AppRegistry.registerComponent('datepicker', () => datepicker);
